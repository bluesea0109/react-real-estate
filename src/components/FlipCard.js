import * as React from 'react';
import { useEffect, useState } from 'react';

const FlipCard = props => {
  const {
    cardStyles: { back, front },
    cardZIndex,
    containerStyle,
    flipDirection,
    flipSpeedFrontToBack,
    flipSpeedBackToFront,
    infinite,
  } = props;

  const [isFlipped, setFlipped] = useState(props.isFlipped);
  const [rotation, setRotation] = useState(0);

  useEffect(() => {
    if (props.isFlipped !== isFlipped) {
      setFlipped(props.isFlipped);
      setRotation(c => c + 180);
    }
  }, [props.isFlipped, isFlipped, setFlipped]);

  const getComponent = key => {
    if (props.children.length !== 2) {
      throw new Error('Component ReactCardFlip requires 2 children to function');
    }
    return props.children[key];
  };

  const frontRotateY = `rotateY(${infinite ? rotation : isFlipped ? 180 : 0}deg)`;
  const backRotateY = `rotateY(${infinite ? rotation + 180 : isFlipped ? 0 : -180}deg)`;
  const frontRotateX = `rotateX(${infinite ? rotation : isFlipped ? 180 : 0}deg)`;
  const backRotateX = `rotateX(${infinite ? rotation + 180 : isFlipped ? 0 : -180}deg)`;

  const styles = {
    back: {
      WebkitBackfaceVisibility: 'hidden',
      backfaceVisibility: 'hidden',
      height: '100%',
      left: '0',
      position: isFlipped ? 'relative' : 'absolute',
      top: '0',
      transform: flipDirection === 'horizontal' ? backRotateY : backRotateX,
      transformStyle: 'preserve-3d',
      transition: `${flipSpeedFrontToBack}s`,
      width: '100%',
      ...back,
    },
    container: {
      perspective: '1000px',
      zIndex: `${cardZIndex}`,
    },
    flipper: {
      height: '100%',
      position: 'relative',
      width: '100%',
    },
    front: {
      WebkitBackfaceVisibility: 'hidden',
      backfaceVisibility: 'hidden',
      height: '100%',
      left: '0',
      position: isFlipped ? 'absolute' : 'relative',
      top: '0',
      transform: flipDirection === 'horizontal' ? frontRotateY : frontRotateX,
      transformStyle: 'preserve-3d',
      transition: `${flipSpeedBackToFront}s`,
      width: '100%',
      zIndex: '2',
      ...front,
    },
  };

  return (
    <div className="card-flip" style={{ ...styles.container, ...containerStyle }}>
      <div className="card-flipper" style={styles.flipper}>
        <div className="card-front" style={styles.front}>
          {getComponent(0)}
        </div>

        <div className="card-back" style={styles.back}>
          {getComponent(1)}
        </div>
      </div>
    </div>
  );
};

FlipCard.defaultProps = {
  cardStyles: {
    back: {},
    front: {},
  },
  cardZIndex: 'auto',
  containerStyle: {},
  flipDirection: 'horizontal',
  flipSpeedBackToFront: 0.6,
  flipSpeedFrontToBack: 0.6,
  infinite: false,
  isFlipped: false,
};

export default FlipCard;
