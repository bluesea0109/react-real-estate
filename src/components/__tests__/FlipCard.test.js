import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from '@testing-library/react';
import 'jest-styled-components';
import FlipCard from '../FlipCard';

it('renders correctly', () => {
  const tree = renderer.create(
    <FlipCard>
      <div></div>
      <div></div>
    </FlipCard>
  );
  expect(tree).toMatchSnapshot();
});

it('Does not render with 0 children', () => {
  let error;
  try {
    shallow(<FlipCard />);
  } catch (e) {
    error = e;
  }
  expect(error).toBeInstanceOf(Error);
});

it('Does not render with 1 child', () => {
  let error;
  try {
    shallow(
      <FlipCard>
        <div></div>
      </FlipCard>
    );
  } catch (e) {
    error = e;
  }
  expect(error).toBeInstanceOf(Error);
});
