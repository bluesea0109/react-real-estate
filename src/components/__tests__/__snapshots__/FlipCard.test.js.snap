// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  style={
    Object {
      "perspective": "1000px",
      "zIndex": "auto",
    }
  }
>
  <div
    className="card-flipper"
    style={
      Object {
        "height": "100%",
        "position": "relative",
        "width": "100%",
      }
    }
  >
    <div
      className="card-front"
      style={
        Object {
          "WebkitBackfaceVisibility": "hidden",
          "backfaceVisibility": "hidden",
          "height": "100%",
          "left": "0",
          "opacity": 1,
          "position": "relative",
          "top": "0",
          "transform": "rotateY(0deg)",
          "transformStyle": "preserve-3d",
          "transition": "0.6s",
          "width": "100%",
          "zIndex": "2",
        }
      }
    >
      <div />
    </div>
    <div
      className="card-back"
      style={
        Object {
          "WebkitBackfaceVisibility": "hidden",
          "backfaceVisibility": "hidden",
          "height": "100%",
          "left": "0",
          "opacity": 0,
          "position": "absolute",
          "top": "0",
          "transform": "rotateY(-180deg)",
          "transformStyle": "preserve-3d",
          "transition": "0.6s",
          "width": "100%",
        }
      }
    >
      <div />
    </div>
  </div>
</div>
`;
